src/posix/routing_manager/router_advertisement.cpp

```diff
@@ -0,0 +1,481 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes implementations for IPv6 Router Advertisement.
+ *
+ */
+
+#define __APPLE_USE_RFC_3542
+
+#include "router_advertisement.hpp"
+
+#if OPENTHREAD_CONFIG_DUCKHORN_BORDER_ROUTER_ENABLE
+
+#include <errno.h>
+#include <netinet/icmp6.h>
+#include <unistd.h>
+
+#include "common/code_utils.hpp"
+#include "common/debug.hpp"
+#include "common/logging.hpp"
+#include "platform/address_utils.hpp"
+
+namespace ot {
+
+namespace Posix {
+
+namespace RouterAdvertisement {
+
+const PrefixInfoOption *RouterAdvMessage::GetNextPrefixInfo(const PrefixInfoOption *aCurPrefixInfo) const
+{
+    const Option *nextOption = aCurPrefixInfo;
+
+    while ((nextOption = GetNextOption(nextOption)) != nullptr)
+    {
+        if (nextOption->GetType() == ND_OPT_PREFIX_INFORMATION)
+        {
+            break;
+        }
+    }
+
+    return reinterpret_cast<const PrefixInfoOption *>(nextOption);
+}
+
+const Option *RouterAdvMessage::GetNextOption(const Option *aCurOption) const
+{
+    const Option *nextOption = nullptr;
+
+    if (aCurOption == nullptr)
+    {
+        nextOption = reinterpret_cast<const Option *>(mOptions);
+    }
+    else
+    {
+        nextOption = aCurOption->GetNextOption();
+    }
+
+    VerifyOrExit(reinterpret_cast<const uint8_t *>(nextOption) >= mOptions, nextOption = nullptr);
+    VerifyOrExit(reinterpret_cast<const uint8_t *>(nextOption) + sizeof(nextOption) <=
+                     reinterpret_cast<const uint8_t *>(this) + GetTotalLength(),
+                 nextOption = nullptr);
+    VerifyOrExit(reinterpret_cast<const uint8_t *>(nextOption->GetNextOption()) <=
+                     reinterpret_cast<const uint8_t *>(this) + GetTotalLength(),
+                 nextOption = nullptr);
+
+exit:
+    return nextOption;
+}
+
+RouterAdvertiser::RouterAdvertiser(RouterSolicitHandler       aRouterSolicitHandler,
+                                   void *                     aRouterSolicitHandlerContext,
+                                   RouterAdvertisementHandler aRouterAdvertisementHandler,
+                                   void *                     aRouterAdvertisementHandlerContext)
+    : mSocketFd(-1)
+    , mRouterSolicitHandler(aRouterSolicitHandler)
+    , mRouterSolicitHandlerContext(aRouterSolicitHandlerContext)
+    , mRouterAdvertisementHandler(aRouterAdvertisementHandler)
+    , mRouterAdvertisementHandlerContext(aRouterAdvertisementHandlerContext)
+{
+}
+
+void RouterAdvertiser::Init()
+{
+    otError             error = OT_ERROR_FAILED;
+    int                 sock;
+    struct icmp6_filter filter;
+    ssize_t             rval;
+    static const int    kTrue  = 1;
+    static const int    kFalse = 0;
+
+    sock = socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6);
+    if (sock < 0)
+    {
+        otLogWarnPlat("failed to open ICMPv6 socket: %s", strerror(errno));
+        ExitNow();
```

Comment:
Should we just exit the program?

Reply:
using `DieNow()` in our latest implementation.